sky/skylet/job_lib.py

```diff
@@ -126,6 +137,77 @@ def colored_str(self):
         return f'{color}{self.value}{colorama.Style.RESET_ALL}'
 
 
+# Only update status of the jobs after this many seconds of job submission,
+# to avoid race condition with `ray job` to make sure it job has been
+# correctly updated.
+# TODO(zhwu): This number should be tuned based on heuristics.
+_PENDING_SUBMIT_TIMEOUT = 5
+
+_PRE_RESOURCE_STATUSES = [JobStatus.PENDING]
+
+
+class JobScheduler:
+    """Base class for job scheduler"""
+
+    def queue(self, job_id: int, cmd: str) -> None:
+        _CURSOR.execute('INSERT INTO pending_jobs VALUES (?,?,?,?)',
+                        (job_id, cmd, 0, int(time.time())))
+        _CONN.commit()
+        set_status(job_id, JobStatus.PENDING)
+        self.schedule_step()
+
+    def remove_job_no_lock(self, job_id: int) -> None:
+        _CURSOR.execute(f'DELETE FROM pending_jobs WHERE job_id={job_id!r}')
+        _CONN.commit()
+
+    def _run_job(self, job_id: int, run_cmd: str):
+        _CURSOR.execute((f'UPDATE pending_jobs SET submit={int(time.time())} '
+                         f'WHERE job_id={job_id!r}'))
+        _CONN.commit()
+        subprocess.Popen(run_cmd, shell=True, stdout=subprocess.DEVNULL)
+
+    def schedule_step(self) -> None:
+        job_owner = getpass.getuser()
+        jobs = self._get_jobs()
+        if len(jobs) > 0:
+            update_status(job_owner)
+        for job_id, run_cmd, submit, _ in jobs:
+            # TODO(mraheja): remove pylint disabling when filelock
+            # version updated
+            # pylint: disable=abstract-class-instantiated
+            with filelock.FileLock(_get_lock_path(job_id)):
+                status = get_status_no_lock(job_id)
+                if status not in _PRE_RESOURCE_STATUSES:
+                    # Job doesn't exist or is running/cancelled
+                    self.remove_job_no_lock(job_id)
+                    continue
+                if submit:
+                    # Next job waiting for resources
+                    return
+                self._run_job(job_id, run_cmd)
+                return
```

Comment:
Should we wait a few seconds and see if the job that is just submitted starts running? If that job starts running, we should continue scheduling the next one.

This is a problem met by `pytest tests/test_smoke.py::test_large_job_queue`, where we canceled multiple jobs with a single `sky cancel` command, and the `cancel` function only call the `schedule_step` once, i.e. only one job is submitted after multiple jobs are cancelled. The remaining PENDING jobs take a long time to be scheduled.

Reply:
Another way to fix this is to have another function called `schedule_bulk_step` to submit all possible PENDING jobs, with the behavior mentioned above, and have the `sky cancel` call the `schedule_bulk_step` instead?

Reply:
Actually, the problem should be resolved if the `schedule_step` is called in the correct place for the next comment https://github.com/skypilot-org/skypilot/pull/1636#discussion_r1218684130